### ISSUE TITLE
example not working

### ISSUE THREAD

**Issue by gurugeek** [longshorej/rustags#1]
First of all thanks for writing another template engine... It looks great!  Unfortunately I just tried your example and I get this error

`error: expected one of `::`, `;`, or `as`, found `*`
  --> src/main.rs:20:20
   |
20 | use rust_tags::tags*;
   |                    ^ expected one of `::`, `;`, or `as` here

error: aborting due to previous error

`
does the lib works only on rust nightly? I am running stable.

**Comment by longshorej** [comment]
Ah, I think that's a typo in the readme. That line should be `use rust_tags::tags::*;`.

Thanks for the interest - let me know if you have any other feedback as it's quite early days for this project. Also check out this other project which uses rust-tags: https://github.com/longshorej/jason-longshore/tree/master/src/pages

**Comment by longshorej** [comment]
Okay - updated the readme with fixed content, verified working with the following:

```
$ rustc -V
rustc 1.22.1 (05e2e1c41 2017-11-22)
-> 0
```

**Comment by gurugeek** [comment]
hi thanks for the quick reply!! now the git examples works WOW awesome! 
FYI the example at 
https://crates.io/crates/rust-tags

is not working ... 

`error[E0433]: failed to resolve. Use of undeclared type or module `tags`
   --> src/main.rs:246:12
    |
246 |     head(&[tags::title(&["My Blog".into()])]),
    |            ^^^^ Use of undeclared type or module `tags`

warning: unused import: `rust_tags::core::*`
  --> src/main.rs:23:5
   |
23 | use rust_tags::core::*;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default`

but glad that the git one does 👍 
for a form I just use  form(&[
right ?

**Comment by longshorej** [comment]
Ah - the crates.io README tracks the latest release instead of `master`. It'll be fixed the next time I do a release.

Yeah, `form(&[...])` will return a `Fragment`. You can place any number of `Fragments` in the slice you pass to `form`. so e.g. 

`form(&[id("my-form"), "this is my <form>".into()])` 

should render as 

`<form id="my-form">this is my &lt;form&gt;</form>`.

**Comment by gurugeek** [comment]
@longshorej thanks again for our prompt reply ! and the text area/field submit have to go inside the fragment too ? would really appreciate a practical example as some things are not so obvious to me! 

I did find out on display a simple variable on my own thou :)   (e.g.  let superman = "superman is here";   and inside the tags  superman.into(),)  It sounds trivial but may be adding it to the example wouldn't hurt !

**Comment by gurugeek** [comment]
one more thing about the css :) 

        head(&[title(&["My Blog".into()])], &[link(&["rel="stylesheet", type="text/css", href="https://cloud.typography.com/7964312/7143592/css/fonts.css"".into()])]),

seems not to work so I think I got the required syntax wrong... will be happy to help with the docs once with some examples once I figure all these bits out !!

**Comment by longshorej** [comment]
Attributes are just function calls that take an `&str`. If an attribute or tag is a keyword, I've simply prepended _ to the name (so e.g. `type` attribute is `fn _type`).

I think you're looking for something like this: `link(&[rel("stylesheet"), _type("text/css"), href("https://cloud.typography.com/7964312/7143592/css/fonts.css")])`

**Comment by longshorej** [comment]
You could even add your own function to abstract this, so something like (note, haven't tested this):

```rust
fn link_css(url: &str) -> Fragment {
  link(&[rel("stylesheet"), _type("text/css"), href(url)])
}
```

and then you can do e.g.

```rust
head(&[
  link_css("my-css1.css"),
  link_css("my-css2.css"),

  ...
])
```

**Comment by gurugeek** [comment]
great @longshorej ! CSS sorted..now what is the tag for 
style="font-family: .... B;font-style: normal;font-weight: 400;font-size:30pt;"
I tried style(&[ but it didn't really work :)

**Comment by gurugeek** [comment]
ah both for form and style I got the same ambiguous issue:

error: `form` is ambiguous
   --> src/main.rs:257:10
    |
257 |          form(&[id("my-form"), "this is my <form>".into()]),
    |          ^^^^
    |
note: `form` could refer to the name imported here
   --> src/main.rs:24:5
    |
24  | use rust_tags::tags::*;
    |     ^^^^^^^^^^^^^^^^^^
note: `form` could also refer to the name imported here
   --> src/main.rs:26:5
    |
26  | use rust_tags::attributes::*;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^
    = note: consider adding an exp

**Comment by longshorej** [comment]
Yeah, I have to think about the best way to handle that still. If you explicitly import the `rust_tags::tags::form` and `rust_tags::attributes::style` that would work, or you can just inline the entire name for these cases, i.e. `rust_tags::tags::form(&[...])`

Your style call is `style("font-family: .... B;font-style: normal;font-weight: 400;font-size:30pt;")`

**Comment by gurugeek** [comment]
great!  @longshorej  ` rust_tags::tags::form(&[id("my-form"), "this is my <form>".into()]),`
and 
`     rust_tags::attributes::style("font-family:Tungsten A, Tungsten B;font-style: normal;font-weight: 400;font-size:30pt;"),`

work fine thank you!
now one more thing (then I should be done!) 
how about textarea, and the button like in normal HTML  

input(type="text" ,name="Title", placeholder="Some text");
                 input (type="submit");
                 textarea (name="Notes", cols="40", rows="5");

I tried 
       `rust_tags::tags::textarea(&[name("notes"), cols ("40"), rows ("5").into()]),`
with no luck 👎

**Comment by longshorej** [comment]
should be something like:

```rust
input(&[_type("text"), name("title"), placeholder("Some text")])

input(&[_type("submit")])

textarea(&[name("notes"), cols("40"), rows("40")])
```

**Comment by gurugeek** [comment]
@longshorej  thanks again! I am really grateful and if you ever set up a donation page let me know as I would be happy to pay back for the great help 👍 

I now have 
`rust_tags::tags::form(&[action("/note"), method("POST"), enctype("text/plain")],),

                input(&[_type("text"), name("title"), placeholder("Some text")]),
                input(&[_type("submit")]),
                textarea(&[name("notes"), cols("40"), rows("40")]),
                                
                rust_tags::tags::form(&[id("my-form"), "this is my <form>".into()]),
`
unfortunately the form tag gets closed so how would I have it closed only after the texture or the submit ? Sorry again if this sounds trivial !

**Comment by longshorej** [comment]
No problem - glad to help, it gives me a better perspective on how I need to build the docs when I get to that.

You'll want all the elements inside the form slice (`&[]`) as well, note that you closed it after the `enctype` call. So move the input and textarea into the form, i.e.

```rust
rust_tags::tags::form(&[action("/note"), method("POST"), enctype("text/plain"),


            input(&[_type("text"), name("title"), placeholder("Some text")]),
            input(&[_type("submit")]),
            textarea(&[name("notes"), cols("40"), rows("40")])
])
```

**Comment by gurugeek** [comment]
works perfectly @longshorej thank you!!
another small thing that is actually important to others too :)  I am trying to fetch some data from a db, just a simple note 

```
let mut content: Option<String> = None;
      
            for row in &db.query("SELECT content FROM notes WHERE id = $1", &[&id]).unwrap() {
                content = Some(row.get(0));
                  }

```

but if I use
` content.into(),`
I get 

```
error[E0277]: the trait bound `rust_tags::core::Fragment: std::convert::From<std::option::Option<std::string::String>>` is not satisfied
   --> src/main.rs:443:25
    |
443 |                 content.into(),
    |                         ^^^^ the trait `std::convert::From<std::option::Option<std::string::String>>` is not implemented for `rust_tags::core::Fragment`
    |
    = help: the following implementations were found:
              <rust_tags::core::Fragment as std::convert::From<i32>>
              <rust_tags::core::Fragment as std::convert::From<&'a str>>
              <rust_tags::core::Fragment as std::convert::From<std::string::String>>
    = note: required because of the requirements on the impl of `std::convert::Into<rust_tags::core::Fragment>` for `std::option::Option<std::string::String>`

error: aborting due to previous error


```
